dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/NeuralNetwork.java

```diff
@@ -185,9 +185,14 @@ public void train(final INDArray input, final int label) {
   public INDArray[] backPropagateTo(final int end,
                                     final INDArray[] activations,
                                     final INDArray label) {
+    if (end == 0) {
+      throw new IllegalArgumentException("The ending layer cannot be the first layer: " +
```

Comment:
The error that the first layer calculates is not used for generating gradients for the first layer. Even though the first layer is a fully connected layer, the errors needed for generating gradients are calculated by the next layer's backpropagation. Thus, the backpropagation for the first layer is unnecessary.
I'll add more comments to this.